server/remote_cache/content_addressable_storage_server/content_addressable_storage_server.go

```diff
@@ -265,12 +284,57 @@ func (s *ContentAddressableStorageServer) BatchReadBlobs(ctx context.Context, re
 			blobRsp.Status = &statuspb.Status{Code: int32(codes.Internal)}
 		} else {
 			blobRsp.Status = &statuspb.Status{Code: int32(codes.OK)}
+			if s.env.GetConfigurator().GetCacheZstdTranscodingEnabled() && clientAcceptsCompressor(req.AcceptableCompressors, repb.Compressor_ZSTD) {
+				blobRsp = zstdCompressResponse(blobRsp, req.AcceptableCompressors)
+			}
 		}
+
 		rsp.Responses = append(rsp.Responses, blobRsp)
 	}
 	return rsp, nil
 }
 
+func (s *ContentAddressableStorageServer) supportsCompressor(compressor repb.Compressor_Value) bool {
+	return compressor == repb.Compressor_IDENTITY ||
+		compressor == repb.Compressor_ZSTD && s.env.GetConfigurator().GetCacheZstdTranscodingEnabled()
+}
+
+func clientAcceptsCompressor(acceptableCompressors []repb.Compressor_Value, compressor repb.Compressor_Value) bool {
+	// Per protocol, IDENTITY is always accepted.
+	if compressor == repb.Compressor_IDENTITY {
+		return true
+	}
+	for _, c := range acceptableCompressors {
+		if c == compressor {
+			return true
+		}
+	}
+	return false
+}
+
+func zstdCompressResponse(blobRsp *repb.BatchReadBlobsResponse_Response, acceptableCompressors []repb.Compressor_Value) *repb.BatchReadBlobsResponse_Response {
+	data, err := zstd.Compress(nil, blobRsp.Data)
+	if err != nil {
+		log.Errorf("Failed to compress blob: %s", err)
```

Comment:
nit: do you really want to log this? It doesn't seem particularly actionable post-development
```suggestion
		log.Errorf("Failed to compress blob: %s", err)
```

Reply:
Since this is never really expected (we should always be able to compress a stream of bytes) an error here could indicate that our compression impl could somehow be broken for some cases we aren't aware of, and users might not tell us about it. We can remove it later if we find that there are cases where it logs errors that seem reasonable / expected, but for now I wanted to be a bit cautious and log them as errors